text/0000-update-version-handling.md

```diff
@@ -0,0 +1,225 @@
+- Feature Name: update_version_handling
+- Start Date: 2019-04-23
+- RFC PR: (leave this empty)
+- Notion Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Change all version-specifying commands to use the `<tool>[@<version>]` syntax from `npm install` and `yarn add`.
+
+# Motivation
+[motivation]: #motivation
+
+Users may regularly want to add multiple tools to their toolchain at the same time. They are also accustomed to being able to install multiple dependencies simultaneously with the `npm install` or `yarn add` commands, both in a package and globally.
+
+- **Local installs:**
+
+    ```sh
+    npm install ember-cli@3.9.1 typescript@3
+    ```
+    
+    ```sh
+    yarn global ember-cli@3.9.1 typescript@3
+    ```
+
+- **Global installs:**
+
+    ```sh
+    npm install --global ember-cli@3.9.1 typescript@3
+    ```
+    
+    ```sh
+    yarn global add ember-cli@3.9.1 typescript@3
+    ```
+
+Notion currently does not support this, requiring individual invocations for each tool:
+
+- **User toolchain:**
+
+    ```sh
+    notion install ember-cli 3.9.1
+    notion install typescript 3
+    ```
+
+- **Project toolchain:**
+
+    ```sh
+    notion pin ember-cli 3.9.1
+    notion pin typescript 3
+    ```
+
+- **Tool retrieval:**
+
+    ```sh
+    notion fetch ember-cli 3.9.1
+    notion fetch typescript 3
+    ```
+
+This is a surprising difference for users coming from `npm` or `yarn`. Switching over to use the same invocation as the existing tools will make adoption smoother and less surprising.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+Installation, fetching, and pinning function much as they do today, with two changes:
+
+- The tool + version combination is now a single string, in the format `tool[@<version>]`. That is, a user may supply a tool with its name only, e.g. `node`, or with a specific version using the `@` to separate the name from the version, e.g. `node@10`, `node@10.5`, `node@10.5.2`, or `node@latest`.
```

Comment:
This RFC uses "version", but would it be more accurate to say "version specifier", or maybe "version range"?

These invocations are mentioned in the doc:

```
node
node@latest
node@1.2.3
node@1.2
```

It is my intuition that it is also possible to do these kinds of things, which are supported by `npm install` and `yarn add`:

```
node@10.1*
node@^10.1
node@~10.1
node@>=10
node@<10
```

These operations are already supported by `jetson install <package> <versionspec>` (through the use of the `semver` crate), but it may be worth spelling out more concretely what is and is not supported here.

For instance there are also other range operators like `1.2.3 - 3.4.5`, `1.2 || 2.2`, etc. that _could_ be supported in the existing `notion install <package> <versionspec>` case, but would probably be too complicated to parse for this change.

Reply:
This is a great call-out – and when I started poking at the implementation, I realized that "version" and "version spec" are different. I'll update it concretely to specify what we *don't* support here as well. I think we'll support versions and version specs, but not range operators (I'll add those to the <b>Unresolved Questions</b> as a thing I think we should defer – probably permanently).

Reply:
Interestingly, in the current implementation, we actually do support all of the operators, including the range and `||` operators. The only trick is that it has to be parsed as a single argument, so you need to either quote it or remove all the whitespace. But I just tried `notion install node "8 || 9"` and it correctly found and installed Node 9.

Reply:
@charlespierce interesting. I *think* we should still be able to support that, with precisely the same restrictions: `notion install "node@8 || 9"`. (I don't expect it really ever to come up, but maintaining support doesn't hurt us.)

Reply:
Updated with this substantially fleshed out (d1ce7fa6); thanks for highlighting it!